Add `invoke_script` to commands to allow for pipelining of scripts

Dav1dde opened the pull request:
Currently it's not possible to add a script to a pipeline, adds a new method `script` to the pipelines.

I also exposed `eval_cmd` and `load_cmd` in the script invocation.

nihohit commented on redis/src/script.rs:
```diff
@@ -206,12 +206,25 @@ impl<'a> ScriptInvocation<'a> {
 
-    fn load_cmd(&self) -> Cmd {
+    /// Returns a command to load the script.
+    pub fn load_cmd(&self) -> Cmd {
```
why does this need to be a public method? the user has the actual script, this uses a single argument from the whole invocation.

Dav1dde replied:
I decided to make both methods public `load_cmd` and `eval_cmd` so they can be independently added to pipelines just through `add_command`. I can change them to be `pub(crate)` if you don't want them to be exposed.

In theory the `load_cmd` could also be a method of `Script`, I kept it here because all the execution (e.g. `load()`) is on the `ScriptInvocation`.

Dav1dde replied:
It can be still loaded by using the `ScriptInvocation`, currently using it like this:

```rs
    pub fn invoke(&self, con: &mut Connection<'_>) -> Result<Vec<CardinalityScriptResult>> {
        match self.pipe.query(con) {
            Ok(result) => Ok(result),
            Err(err) if err.kind() == redis::ErrorKind::NoScriptError => {
                self.script.load_redis(con)?;
                self.pipe
                    .query(con)
                    .map_err(relay_redis::RedisError::Redis)
                    .map_err(Into::into)
            }
            Err(err) => Err(relay_redis::RedisError::Redis(err).into()),
        }
    }
```

But it can't be added to the pipeline without it, but that can also be solved by adding another fn to the `implement_commands`.

nihohit commented on redis/src/pipeline.rs:
```diff
@@ -293,6 +293,16 @@ macro_rules! implement_pipeline_commands {
+            ///
+            /// Shorthand for `add_command(invocation.eval_cmd())`.
+            #[inline]
+            #[cfg(feature = "script")]
```
instead of adding it here, can you add it to the `implement_commands` macro invocation in commands/mod.rs? that way it will be added to multiple relevant objets.

Dav1dde replied:
Sounds, good. I didn't initially add it to `implement_commands` because it is not an actual Redis command.

nihohit reviewed (commented):
please add tests & public documentation.+

Dav1dde commented:
@nihohit Changed some stuff:

- Moved the existing tests together with the new ones into a separate file. 
- Kept the the `load_cmd` and `eval_cmd` methods private.
- Moved the `script` fn to the `implement_commands`.
- Updated the lib docs with an example.

nihohit commented on redis/tests/test_script.rs:
```diff
@@ -0,0 +1,70 @@
+    let ctx = TestContext::new();
+    let mut con = ctx.connection();
+
+    redis::cmd("SCRIPT").arg("FLUSH").execute(&mut con);
```
this is unnecessary - `TestContext` creates a new server.

nihohit commented on redis/tests/test_script.rs:
```diff
@@ -0,0 +1,68 @@
+
+#[test]
+#[cfg(feature = "script")]
+fn test_script_pipeline_no_autoload() {
```
test_script_that_is_not_loaded_fails_on_pipeline_invocation ?

nihohit reviewed (approved):
@jaymell WDYT?

Dav1dde commented:
Thanks for the review, updated with your comments and found some small doc issues I also fixed.

I am a bit unsure about the `script` function name since it does collide with [`SCRIPT`](https://redis.io/commands/script/).

nihohit commented:
> I am a bit unsure about the script function name since it does collide with [SCRIPT](https://redis.io/commands/script/).

maybe `invoke_script`?

nihohit commented:
@Dav1dde can you please rebase?

Dav1dde commented:
@nihohit squashed all the commits and rebased to latest main
